localization/util/pose_initializer/src/pose_initializer_core.cpp

```diff
@@ -188,26 +192,33 @@ bool PoseInitializer::getHeight(
 }
 
 bool PoseInitializer::callAlignService(
-  const geometry_msgs::PoseWithCovarianceStamped & input_pose_msg,
-  const geometry_msgs::PoseWithCovarianceStamped::Ptr & output_pose_msg_ptr)
+  const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr input_pose_msg,
+  const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr output_pose_msg_ptr)
 {
-  autoware_localization_srvs::PoseWithCovarianceStamped srv;
-  srv.request.pose_with_cov = input_pose_msg;
-
-  ROS_INFO("[pose_initializer] call NDT Align Server");
-  if (ndt_client_.call(srv)) {
-    ROS_INFO("[pose_initializer] called NDT Align Server");
+  auto req =
+    std::make_shared<autoware_localization_srvs::srv::PoseWithCovarianceStamped::Request>();
+  req->pose_with_cov = *input_pose_msg;
+
+  RCLCPP_INFO(get_logger(), "call NDT Align Server");
+
+  auto result = ndt_client_->async_send_request(req);
+  // Wait for the result. `shared_from_this` is fine since this class is only used with `make_shared`.
+  if (
+    rclcpp::spin_until_future_complete(this->shared_from_this(), result) ==
```

Comment:
We had a problem with Autoware.Auto where we couldn't do spin_until_future_complete() inside a Node class.
The node itself is already added to the executor in spin() function inside the main function, and spin_until_future_complete also tries to add the node to executor inside the function. Since executor throws error when the user tries to add same node twice, it will abort when process reach to this line.

Reply:
Workaround is to call async_send_request with a callback, or handover the executor to the PoseInitializer and do the similar thing that is done in spin_until_future_complete() function by ourselves.

Reply:
We might want to discuss about this in the meeting on Monday.

Reply:
Thanks for pointing that out! How about just calling `wait()` or `get()` on the returned future? I can't imagine that interfering with the node, or the design would be broken.

Reply:
Okay, those just block forever. I rewrote it with a callback, which was fortunately easy. The drawback is that there is no way to detect failure there: If the service is offline, the callback is just never called. There's no good way to set a timeout either.

I wonder if `.wait()` would work with a `MultiThreadedExecutor`.

Reply:
It might work with MultiThreadedExecutor, but I think it would cause larger impact since we would then have to think about using mutex::lock in callbacks.

Reply:
If you prefer, we could at least have a minimal check like this:
* have a member variable that identifies the latest response received and one for the latest request sent
* set the response identifier in the callback
* before the next client call, check that the response identifier corresponds to the request identifier, and then set the request variable

Reply:
I see. that sounds good to me.
Since it is now a callback and does not going to block the process, we probably have to make sure that we don't send new service request before previous request is done.

Reply:
Done. I added a sequence number to the request, and in the ndt_scan_matcher I will need to copy that number from the request into the response.

Reply:
Thanks. Looks good to me.